lib/std/debug.zig

```diff
@@ -50,163 +40,378 @@ pub const LineInfo = struct {
+                    if (byte == '\n') {
+                        return;
+                    }
+                }
+                if (byte == '\n') {
+                    line += 1;
+                    column = 1;
+                } else {
+                    column += 1;
+                }
+            }
+
+            if (amt_read < buf.len) return error.EndOfFile;
+        }
+    }
+
+    pub fn attemptWriteLineFromSourceFile(writer: anytype, line_info: LineInfo) !bool {
+        writeLineFromFileAnyOs(writer, line_info) catch |err| {
+            switch (err) {
+                error.EndOfFile, error.FileNotFound => {},
+                error.BadPathName => {},
+                error.AccessDenied => {},
+                else => return err,
+            }
+            return false;
+        };
+        return true;
+    }
+
+    /// When overriding panicTerminate, if you *must* lock a mutex, be careful not to panic!
+    /// Otherwise, this will would call panicTerminate recursively and deadlock.
+    pub const panicTerminate = os.abort;
+};
+
+const is_default_config = @hasDecl(root, "debug_config");
+const config = if (is_default_config)
+    root.debug_config
+else
+    default_config;
+
+const has_writer_decl = @hasDecl(config, "getWriter");
+
+/// The function used to get a writer for debugging.
+/// Slightly different name than in config to avoid redefinition in default.
+const getWrit = if (has_writer_decl)
+    config.getWriter
+else
+    default_config.getWriter;
+
+/// The function used to determine which escape codes can be used.
+/// Slightly different name than in config to avoid redefinition in default.
+const getTTYConfig = if (@hasDecl(config, "detectTTYConfig"))
+    config.detectTTYConfig
+else if (has_writer_decl)
+    @compileError("getWriter exists in config, so detectTTYConfig must also exist")
+else
+    default_config.detectTTYConfig;
+
+const has_address_symbol_mapping_decl = @hasDecl(config, "AddressSymbolMapping");
+const using_default_address_symbol_mapping = is_default_config or
+    !has_address_symbol_mapping_decl;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const AddrSymMapping = if (has_address_symbol_mapping_decl)
+    config.AddressSymbolMapping
+else
+    default_config.AddressSymbolMapping;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const writeLineFromSourceFile = if (@hasDecl(config, "attemptWriteLineFromSourceFile"))
+    config.attemptWriteLineFromSourceFile
+else
+    default_config.attemptWriteLineFromSourceFile;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const panicTerm = if (@hasDecl(config, "panicTerminate"))
+    config.panicTerminate
+else
+    default_config.panicTerminate;
+
+var mutex = std.Thread.Mutex{};
+
+const HeldWriter = HeldValue(@TypeOf(getWrit()));
+pub fn heldWriter() HeldWriter {
+    return .{ .value = getWrit(), .held = mutex.acquire() };
+}
```

Comment:
The PR makes a `pub` function for acquiring the writer used by `debug` in a locked state (log has something similar). Related Questions:
 - Is there any reason to allow access to this writer outside of `std.debug`? I think so?
 - Is there any case where we would want to allow access to the writer from outside from file without also acquiring? I think no - this seems like a footgun.

Reply:
> Is there any reason to allow access to this writer outside of std.debug? I think so?

This writer is used only by the debug interface and it'd be better to hide it from the public API to avoid unexpected deadlocks.

> Is there any case where we would want to allow access to the writer from outside from file without also acquiring?

Keep in mind that if something panics during a panic we should be able to write out something, see #8058 for more info.

Reply:
> > Is there any reason to allow access to this writer outside of std.debug? I think so?
> 
> This writer is used only by the debug interface and it'd be better to hide it from the public API to avoid unexpected deadlocks.

Makes sense. I will remove this (and the equivalent log function)

> 
> > Is there any case where we would want to allow access to the writer from outside from file without also acquiring?
> 
> Keep in mind that if something panics during a panic we should be able to write out something, see #8058 for more info.

Yep, I am aware of this. I added a comment about this here: https://github.com/rgreenblatt/zig/blob/c8856e57b7782d2447dc695fa67705c31373aeb5/lib/std/debug.zig#L442-L446